### PR TITLE
Add initial conditions for FV variables

### LINKED ISSUE #25824: Create initial condition system for finite volume variables.
## Reason
The current initial condition system relies heavily on the `Assembly` and `VariableData` machanisms which is not used in FV. This would simplify code and allow more flexibility. 
Furthermore, we need this as a prereq for #25722 .

## Design
Add new system and thread.

## Impact
Will allow flexibility for linear system assemblies and faster finite volume implementations.

### PR DESCRIPTION
**Pull request by grmnptr**
Closes #25824

### CODE REVIEW

File: framework/include/fvics/FVConstantIC.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+class InputParameters;
+
+namespace libMesh
+{
+class Point;
+}
+
+template <typename T>
+InputParameters validParams();
+
```
Comment (lindsayad) [inline framework/include/fvics/FVConstantIC.h:23]: Remove all of these

File: framework/include/fvics/FVInitialConditionBase.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  FVInitialConditionBase(const InputParameters & parameters);
+
+  virtual ~FVInitialConditionBase();
+
+  static InputParameters validParams();
+
+  /**
+   * retrieves the MOOSE variable that this initial condition acts upon
+   */
+  virtual MooseVariableFEBase & variable() = 0;
```
Comment (lindsayad) [inline framework/include/fvics/FVInitialConditionBase.h:54]: Use a different typedef

File: framework/include/mesh/ElemInfo.h [hunk 3]
```diff
@@ -34,6 +34,16 @@ class ElemInfo
   const Elem * elem() const { return _elem; }
   Real volume() const { return _volume; }
   const Point & centroid() const { return _centroid; }
+  Real coordFactor() const { return _coord_transform_factor; }
+  Real & coordFactor() { return _coord_transform_factor; }
+  const std::vector<std::vector<dof_id_type>> & dofID() const { return _dof_ids; }
+  std::vector<std::vector<dof_id_type>> & dofID() { return _dof_ids; }
+
+  void setDofIDs(const std::vector<std::vector<dof_id_type>> & dof_ids)
```
Comment (lindsayad) [inline framework/include/mesh/ElemInfo.h:42]: I would use indices instead of ids. In libMesh jargon, dof object ids are the ids of nodes/elements. Indices are used for indices in the DofMap

File: framework/include/mesh/MooseMesh.h [hunk 4]
```diff
@@ -1149,14 +1160,16 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
    * Cache \p elem and \p neighbor dof indices information for variables in all the local \p
    * FaceInfo objects to save computational expense
    */
-  void cacheVarIndicesByFace(const std::vector<const MooseVariableFieldBase *> & moose_vars);
+  void cacheVarIndicesByFace() const;
```
Comment (lindsayad) [inline framework/include/mesh/MooseMesh.h:1163]: we should probably rename this method as it doesn't cache indices any more

File: framework/include/problems/DumpObjectsProblem.h [hunk 5]
```diff
@@ -83,6 +83,7 @@ class DumpObjectsProblem : public FEProblemBase
   captureDump(addAuxScalarKernel,     "AuxScalarKernels")
   captureDump(addAuxVariable,         "AuxVariables")
   captureDump(addBoundaryCondition,   "BCs")
+  captureDump(addFVBC,  "FVBCs")
```
Comment (lindsayad) [inline framework/include/problems/DumpObjectsProblem.h:86]: indentation

File: framework/include/systems/SystemBase.h [hunk 6]
```diff
@@ -543,6 +543,18 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
...
+   * Get the number of field variables in this system
+   * @return the number of field variables
+   */
+  virtual unsigned int nFieldVariables() const;
+
+  /**
+   * Get the number of finite volume variables in this system
+   * @return the number of finite volume variables
+   */
+  virtual unsigned int nFVVariables() const;
```
Comment (lindsayad) [inline framework/include/systems/SystemBase.h:556]: seems like maybe these don't need to be virtual

File: framework/src/auxkernels/AuxKernel.C [hunk 7]
```diff
@@ -179,6 +179,16 @@ AuxKernelTempl<ComputeValueType>::AuxKernelTempl(const InputParameters & paramet
     paramError("variable",
                "Variable family " + Moose::stringify(type.family) + " is not supported at order " +
                    Moose::stringify(type.order) + " by the AuxKernel system.");
+
+  // The AuxKernel computation loops rely on quadrature point evaluations. Up until the introduction
+  // of FVAuxKernels we will require quadrature point evaluations for the variables general
+  // AuxKernels act on.
```
Comment (lindsayad) [inline framework/src/auxkernels/AuxKernel.C:185]: Based on this comment, I'm not convinced of the need for calling `requireQpComputations()` since the aux kernel will never call `_u[_qp]` itself

File: framework/src/base/Moose.C [hunk 8]
```diff
@@ -146,6 +146,9 @@ addActionTypes(Syntax & syntax)
   registerMooseObjectTask("add_ic",                       InitialCondition,       false);
   appendMooseObjectTask  ("add_ic",                       ScalarInitialCondition);
 
+  registerMooseObjectTask("add_fv_ic",                    FVInitialCondition,     false);
+  appendMooseObjectTask  ("add_fv_ic",                    ScalarInitialCondition);
```
Comment (lindsayad) [inline framework/src/base/Moose.C:150]: I'm not quite sure what this does. Does this allow building initial conditions for scalar variables in the `FVICs` syntax?

File: framework/src/base/Moose.C [hunk 9]
```diff
@@ -310,7 +313,7 @@ addActionTypes(Syntax & syntax)
                            "(setup_function_complete)"
                            "(setup_adaptivity)"
                            "(set_adaptivity_options)"
-                           "(add_ic)"
+                           "(add_ic add_fv_ic)"
```
Comment (lindsayad) [inline framework/src/base/Moose.C:316]:
```suggestion
                           "(add_ic, add_fv_ic)"
```

File: framework/src/fvics/FVInitialConditionWarehouse.C [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    ic->initialSetup();
+}
+
+void
+FVInitialConditionWarehouse::addObject(std::shared_ptr<FVInitialConditionBase> object,
+                                       THREAD_ID tid,
+                                       bool recurse)
+{
+  // Check that when object is boundary restricted that the variable is nodal
+  const MooseVariableFEBase & var = object->variable();
```
Comment (lindsayad) [inline framework/src/fvics/FVInitialConditionWarehouse.C:35]:
```suggestion
  const auto & var = object->variable();
```

File: framework/src/fvics/FVInitialConditionWarehouse.C [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+  // Block Restricted
+  if (object->blockRestricted())
+  {
+    std::map<std::string, std::set<SubdomainID>>::const_iterator iter =
```
Comment (lindsayad) [inline framework/src/fvics/FVInitialConditionWarehouse.C:40]:
```suggestion
    auto iter =
```

File: framework/src/fvics/FVInitialConditionWarehouse.C [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+      mooseError("The initial condition '",
+                 object->name(),
+                 "' is being defined on a block that already has an initial condition defined.");
+    else
+      _block_ics[tid][var.name()].insert(object->blockIDs().begin(), object->blockIDs().end());
+  }
+  // Non-restricted
+  else
+  {
+    std::map<std::string, std::set<SubdomainID>>::const_iterator iter =
```
Comment (lindsayad) [inline framework/src/fvics/FVInitialConditionWarehouse.C:54]:
```suggestion
    auto iter =
```

File: framework/src/mesh/MooseMesh.C [hunk 13]
```diff
@@ -3339,6 +3356,8 @@ MooseMesh::buildFiniteVolumeInfo() const
   if (!_finite_volume_info_dirty)
     return;
 
+  // mooseError("Lemme see the stack");
+
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:3360]: remove

File: framework/src/mesh/MooseMesh.C [hunk 14]
```diff
@@ -3445,13 +3467,21 @@ MooseMesh::buildFiniteVolumeInfo() const
...
   }
+
+  for (auto & ei : _elem_to_elem_info)
+    if (ei.second.elem()->processor_id() == this->processor_id())
+      _elem_info.push_back(&ei.second);
+
+  computeFiniteVolumeCoords();
+
+  cacheVarIndicesByFace();
+
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C]:
```suggestion
  cacheVarIndicesByFace();
```

File: framework/src/mesh/MooseMesh.C [hunk 15]
```diff
@@ -3568,6 +3615,63 @@ MooseMesh::cacheVarIndicesByFace(const std::vector<const MooseVariableFieldBase
...
+void
+MooseMesh::cacheFVElementalDoFs() const
+{
+  const unsigned int num_eqs = _app.feProblem().es().n_systems();
+
+  for (auto & elem_info_pair : _elem_to_elem_info)
+  {
+    ElemInfo & elem_info = elem_info_pair.second;
+    std::vector<std::vector<dof_id_type>> dof_vector;
+    // dof_vector.clear();
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:3623]:
```suggestion
```

File: framework/src/mesh/MooseMesh.C [hunk 16]
```diff
@@ -3568,6 +3615,63 @@ MooseMesh::cacheVarIndicesByFace(const std::vector<const MooseVariableFieldBase
...
+    for (const auto i : make_range(_app.feProblem().numNonlinearSystems()))
+      if (_app.feProblem().getNonlinearSystemBase(i).nFVVariables())
+      {
+        auto & sys = _app.feProblem().getNonlinearSystemBase(i);
+        dof_vector[sys.number()].clear();
+        dof_vector[sys.number()].resize(sys.nVariables(), dof_id_type(0));
+        const auto & variables = sys.getVariables(0);
+        for (const auto & var : variables)
+        {
+          const std::set<SubdomainID> & var_subdomains = var->blockIDs();
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:3635]:
```suggestion
          const auto & var_subdomains = var->blockIDs();
```

File: framework/src/mesh/MooseMesh.C [hunk 17]
```diff
@@ -3568,6 +3615,63 @@ MooseMesh::cacheVarIndicesByFace(const std::vector<const MooseVariableFieldBase
...
+      }
+
+    if (_app.feProblem().getAuxiliarySystem().nFVVariables())
+    {
+      auto & sys = _app.feProblem().getAuxiliarySystem();
+      dof_vector[sys.number()].resize(sys.nVariables(), libMesh::DofObject::invalid_id);
+      const auto & aux_variables = sys.getVariables(0);
+      for (const auto & var : aux_variables)
+      {
+        const std::set<SubdomainID> & var_subdomains = var->blockIDs();
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:3655]:
```suggestion
        const auto & var_subdomains = var->blockIDs();
```

File: framework/src/problems/DisplacedProblem.C [hunk 18]
```diff
@@ -278,7 +281,11 @@ DisplacedProblem::updateMesh(bool mesh_changing)
   _mesh.getMesh().clear_point_locator();
 
   // The mesh has changed. Face information normals, areas, etc. must be re-calculated
-  _mesh.finiteVolumeInfoDirty();
+  if (haveFV())
+  {
+    _mesh.finiteVolumeInfoDirty();
+    _mesh.buildFiniteVolumeInfo();
```
Comment (lindsayad) [inline framework/src/problems/DisplacedProblem.C:291]: Usually this just got rebuilt as needed by other calls but I suppose it's fine to call it explicitly

File: framework/src/problems/FEProblemBase.C [hunk 19]
```diff
@@ -5623,6 +5683,11 @@ FEProblemBase::init()
     es().init();
   }
 
+  // Now that the equation system and the dof distribution is done, we can generate the
+  // finite volume-related parts if needed.
+  if (haveFV())
+    _mesh.buildFiniteVolumeInfo();
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C]: it would be good for me to understand why these new explicit calls are needed

File: framework/src/systems/SystemBase.C [hunk 20]
```diff
@@ -1193,9 +1212,6 @@ SystemBase::update(const bool update_libmesh_system)
 {
   if (update_libmesh_system)
     system().update();
-  std::vector<VariableName> std_field_variables;
-  getStandardFieldVariableNames(std_field_variables);
-  cacheVarIndicesByFace(std_field_variables);
```
Comment (lindsayad) [inline framework/src/systems/SystemBase.C:1198]: I suppose maybe this is why

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/problems/DisplacedProblem.C (hunk 18 above)
Comment (grmnptr) [reply framework/src/problems/DisplacedProblem.C:291]: Yeah I am removing the calls from accessors.

**Comment by lindsayad** [comment]
I will unsubscribe for now. You can mention me when this is moved out of draft

File: framework/include/mesh/ElemInfo.h (hunk 3 above)
Comment (grmnptr) [reply framework/include/mesh/ElemInfo.h:42]: Gotcha

File: framework/src/base/Moose.C (hunk 8 above)
Comment (grmnptr) [reply framework/src/base/Moose.C:150]: Ahhh great catch!

File: framework/src/auxkernels/AuxKernel.C (hunk 7 above)
Comment (grmnptr) [reply framework/src/auxkernels/AuxKernel.C:185]: This is needed because in the AuxKernel we have this:
```
_var.setDofValues(_local_sol);
```
which operates both on `_vector_tags_dof_u` and `_vector_tag_u` which are resized in `MooseVariableDataFV`
```
template <typename OutputType>
void
MooseVariableDataFV<OutputType>::prepareIC()
{
  // TODO: implement this function
  initDofIndices();
  _vector_tags_dof_u[_solution_tag].resize(_dof_indices.size());

  mooseAssert(_qrule, "We should have a non-null qrule");
  const auto nqp = _qrule->n_points();
  _vector_tag_u[_solution_tag].resize(nqp);
}
```
but in the FV initial condition loop we don't want to use the data variables. 

Now that I think about it, I can just call the `prepareIC` functions from the loop and just have it as an empty function override in the Linear variable which is going to come.

File: framework/src/problems/DisplacedProblem.C (hunk 18 above)
Comment (grmnptr) [reply framework/src/problems/DisplacedProblem.C:292]: My reasoning is that I would not like the mix the functionalities. I extended the assert in the accessor when the lookup doesn't find anything.

File: framework/src/auxkernels/AuxKernel.C (hunk 7 above)
Comment (grmnptr) [reply framework/src/auxkernels/AuxKernel.C:185]: Ok that does the trick too. Will go with that since it is much closer to the original approach.

Comment (grmnptr) [reply framework/src/auxkernels/AuxKernel.C]: Yeah never mind, I don't know what I was thinking. We have a loop over `ElemInfos` here without resetting anything in the assembly/data members. So the `prepareIC` approach does not work.

Comment (grmnptr) [reply framework/src/auxkernels/AuxKernel.C]: Well, in a hacky way it still seems to work. The data members are not used, they are just resized so that they can be populated by the auxkernels.

File: framework/doc/content/source/fvics/FVFunctionIC.md [hunk 21]
```diff
@@ -0,0 +1,17 @@
+# FVFunctionIC
```
Comment (lindsayad) [inline framework/doc/content/source/fvics/FVFunctionIC.md:1]: It might make sense to just have a `FVFunctorIC`

File: framework/doc/content/source/fvics/FVInitialCondition.md [hunk 22]
```diff
@@ -0,0 +1,3 @@
+# FVInitialCondition
```
Comment (lindsayad) [inline framework/doc/content/source/fvics/FVInitialCondition.md:1]: Is this file needed?

File: framework/doc/content/syntax/FVICs/index.md [hunk 23]
```diff
@@ -0,0 +1,33 @@
+# FVICs System
+
+## Description
+
+The `FVICs` block within an input file is utilized to define the initial (starting) conditions for
```
Comment (lindsayad) [inline framework/doc/content/syntax/FVICs/index.md:5]:
```suggestion
The `FVICs` block within an input file is used to define the initial (starting) conditions for
```

File: framework/include/fvics/FVInitialConditionBase.h [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * description
+ */
+class FVInitialConditionBase : public MooseObject,
+                               public BlockRestrictable,
+                               public FunctionInterface,
+                               public Restartable,
+                               public DependencyResolverInterface,
+                               public ADFunctorInterface
```
Comment (lindsayad) [inline framework/include/fvics/FVInitialConditionBase.h:37]: I think this can use the `NonADFunctorInterface`

File: framework/include/fvics/FVInitialConditionBase.h [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+
+  static InputParameters validParams();
+
+  /**
+   * retrieves the MOOSE variable that this initial condition acts upon
+   */
+  virtual MooseVariableFieldBase & variable() = 0;
+
+  /**
+   * Workhorse method for projecting the initial conditions for boundary restricted initial
```
Comment (lindsayad) [inline framework/include/fvics/FVInitialConditionBase.h:57]: Only boundary restricted?

File: framework/include/fvics/FVInitialConditionBase.h [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+   * conditions
+   */
+  virtual void computeElement(const ElemInfo & elem_info) = 0;
+
+  /**
+   * Gets called at the beginning of the simulation before this object is asked to do its job.
+   * Note: This method is normally inherited from SetupInterface.  However in this case it makes
+   * no sense to inherit the other virtuals available in that class so we are adding this virtual
+   * directly to this class with out the extra inheritance.
```
Comment (lindsayad) [inline framework/include/fvics/FVInitialConditionBase.h:66]:
```suggestion
   * directly to this class without the extra inheritance.
```

File: framework/include/mesh/ElemInfo.h [hunk 27]
```diff
@@ -45,4 +49,13 @@ class ElemInfo
   Real _volume;
   /// Centroid of the element
   Point _centroid;
+  /// Cached coordinate transformation factor
+  Real _coord_transform_factor;
+  /// Cached dof id mainly for segregated linear FV evaluations
```
Comment (lindsayad) [inline framework/include/mesh/ElemInfo.h:54]:
```suggestion
  /// Cached dof indices mainly for segregated linear FV evaluations
```

File: framework/include/mesh/MooseMesh.h [hunk 28]
```diff
@@ -1146,17 +1157,18 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
...
 
   /**
-   * Cache \p elem and \p neighbor dof indices information for variables in all the local \p
-   * FaceInfo objects to save computational expense
+   * Cache if variables live on the elements connected by the FaceInfo objects
    */
-  void cacheVarIndicesByFace(const std::vector<const MooseVariableFieldBase *> & moose_vars);
+  void cacheFaceInfoVariableOwnership() const;
+
+  void cacheFVElementalDoFs() const;
```
Comment (lindsayad) [inline framework/include/mesh/MooseMesh.h:1164]: doxygen

File: framework/include/mesh/MooseMesh.h [hunk 29]
```diff
@@ -1252,7 +1264,12 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
   /**
    * Mark the finite volume information as dirty
    */
-  void finiteVolumeInfoDirty() { _finite_volume_info_dirty = true; }
+  void markFiniteVolumeInfoDirty() { _finite_volume_info_dirty = true; }
+
+  /**
+   * Return if the finite volume information is dirty
```
Comment (lindsayad) [inline framework/include/mesh/MooseMesh.h:1270]:
```suggestion
   * @return whether the finite volume information is dirty
```

File: framework/src/fvics/FVInitialConditionTempl.C [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+  {
+    // This line is supposed to throw an error when the user tries to compile this function with
+    // types that are not supported. This is the reason we needed the always_false function. Hope as
+    // C++ gets nicer, we can do this in a nicer way.
+    static_assert(Moose::always_false<T>,
+                  "Initial conditions not is not implemented for the used type!");
+  }
+}
+
+template class FVInitialConditionTempl<Real>;
```
Comment (lindsayad) [inline framework/src/fvics/FVInitialConditionTempl.C:65]: I think it will be impossible to trigger this assertion as you are defining this in the .C file and thus you control its instantiations. But I think this is still fine as it helps convey intent to a developer who looks at this file

File: framework/src/fvics/FVInitialConditionWarehouse.C [hunk 31]
```diff
@@ -0,0 +1,64 @@
...
+  // Non-restricted
+  else
+  {
+    auto iter = _block_ics[tid].find(var.name());
+    if (iter != _block_ics[tid].end())
+      mooseError("The initial condition '",
+                 object->name(),
+                 "' is being defined on a block that already has an initial condition defined.");
+    else
+      _block_ics[tid][var.name()].insert(Moose::ANY_BLOCK_ID);
```
Comment (lindsayad) [inline framework/src/fvics/FVInitialConditionWarehouse.C:59]: You already did the finds. No reason to do the lookups again. Just do `iter->second.insert`

File: framework/include/mesh/MooseMesh.h [hunk 32]
```diff
@@ -1277,6 +1294,19 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
    */
   bool isSplit() const { return _is_split; }
 
+  /**
+   * Builds the face info vector that stores meta-data needed for looping over and doing
+   * calculations based on mesh faces and elements in a finite volume setting.
+   * We also build a vector of elem info objects which cache volumes and centroids for elements.
+   * We build finite volume information only upon request and only if the
+   * \p _finite_volume_info_dirty flag is false, either because this method has yet to be called or
```
Comment (lindsayad) [inline framework/include/mesh/MooseMesh.h:1302]: I think this doxygen needs to be updated

File: framework/src/mesh/MooseMesh.C [hunk 33]
```diff
@@ -3521,9 +3546,22 @@ MooseMesh::deleteRemoteElements()
 }
 
 void
-MooseMesh::cacheVarIndicesByFace(const std::vector<const MooseVariableFieldBase *> & moose_vars)
+MooseMesh::cacheFaceInfoVariableOwnership() const
 {
-  buildFiniteVolumeInfo();
+  std::vector<const MooseVariableFieldBase *> moose_vars;
+
+  for (const auto i : make_range(_app.feProblem().numNonlinearSystems()))
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:3553]:
```suggestion
  mooseAssert(!Threads::in_threads, "Performing writes to elem info dof indices. This must be done unthreaded");

  for (const auto i : make_range(_app.feProblem().numNonlinearSystems()))
```

Comment (lindsayad) [reply framework/src/mesh/MooseMesh.C:3553]: This is probably overly paranoid but I think of this whenever I see hard-coded THREAD_ID arguments

File: framework/src/mesh/MooseMesh.C [hunk 34]
```diff
@@ -3569,6 +3607,62 @@ MooseMesh::cacheVarIndicesByFace(const std::vector<const MooseVariableFieldBase
   }
 }
 
+void
+MooseMesh::cacheFVElementalDoFs() const
+{
+  const unsigned int num_eqs = _app.feProblem().es().n_systems();
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:3613]: I would add a similar thread assertion as I proposed above

File: framework/src/problems/FEProblemBase.C [hunk 35]
```diff
@@ -3038,6 +3048,47 @@ FEProblemBase::addInitialCondition(const std::string & ic_name,
...
+  checkICRestartError(ic_name, name, var_name);
+
+  parameters.set<SubProblem *>("_subproblem") = this;
+
+  // field IC
+  if (hasVariable(var_name))
+  {
+    for (THREAD_ID tid = 0; tid < libMesh::n_threads(); ++tid)
+    {
+      MooseVariableFEBase & var = getVariable(
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:3072]:
```suggestion
      auto & var = getVariable(
```

File: framework/src/problems/FEProblemBase.C [hunk 36]
```diff
@@ -5617,6 +5677,16 @@ FEProblemBase::init()
...
   }
 
+  // Now that the equation system and the dof distribution is done, we can generate the
+  // finite volume-related parts if needed.
+  if (haveFV())
+  {
+    _mesh.buildFiniteVolumeInfo();
+    _mesh.computeFiniteVolumeCoords();
+    _mesh.cacheFaceInfoVariableOwnership();
+    _mesh.cacheFVElementalDoFs();
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:5687]: Would it make sense to have a single `MooseMesh` method that does all four of these calls?

**Review by lindsayad — CHANGES_REQUESTED** [review]
Reviewed framework. Will come back for modules shortly

File: framework/doc/content/source/fvics/FVFunctionIC.md (hunk 21 above)
Comment (GiudGiud) [reply framework/doc/content/source/fvics/FVFunctionIC.md:1]: you need dependency resolution then in case a variable has been passed

**Review by lindsayad — COMMENTED** [review]
turns out there was nothing more to add 😄

File: framework/include/fvics/FVInitialConditionBase.h (hunk 25 above)
Comment (grmnptr) [reply framework/include/fvics/FVInitialConditionBase.h:57]: I wanted to use block :(, no boundary restriction for us at this point

File: framework/src/problems/FEProblemBase.C (hunk 36 above)
Comment (grmnptr) [reply framework/src/problems/FEProblemBase.C:5687]: I can definitely do that! Just wanted to have the flexibility to call them all separately since some unit tests rely only a few of them.

Comment (grmnptr) [reply framework/src/problems/FEProblemBase.C:5687]: This reminds me that I'll need to add a unit test for elemental dof indices.

File: framework/doc/content/source/fvics/FVInitialCondition.md (hunk 22 above)
Comment (grmnptr) [reply framework/doc/content/source/fvics/FVInitialCondition.md:1]: Nope, nice catch!

File: framework/doc/content/source/fvics/FVFunctionIC.md (hunk 21 above)
Comment (lindsayad) [reply framework/doc/content/source/fvics/FVFunctionIC.md:1]: Function works for me then to avoid possible issues

File: framework/src/fvics/FVInitialConditionWarehouse.C (hunk 31 above)
Comment (grmnptr) [reply framework/src/fvics/FVInitialConditionWarehouse.C]: We have 2 keys here. The find is on the second one and we throw an error if we already have it. This means that the line you highlighted is only called if the lookup does not find anything. So in this case we just insert a new key with this call. Otherwise it is a segfault trying to insert something to the `end()`.

File: framework/src/mesh/MooseMesh.C [hunk 37]
```diff
@@ -3507,8 +3507,10 @@ MooseMesh::computeFiniteVolumeCoords() const
   }
 
   for (auto & ei : _elem_to_elem_info)
+  {
     coordTransformFactor(
         *this, ei.second.subdomain_id(), ei.second.centroid(), ei.second.coordFactor());
+  }
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:3513]: No need for braces

File: framework/src/problems/DisplacedProblem.C [hunk 38]
```diff
@@ -289,10 +284,7 @@ DisplacedProblem::updateMesh(bool mesh_changing)
   if (haveFV())
   {
     _mesh.markFiniteVolumeInfoDirty();
-    _mesh.buildFiniteVolumeInfo();
-    _mesh.computeFiniteVolumeCoords();
-    _mesh.cacheFaceInfoVariableOwnership();
-    _mesh.cacheFVElementalDoFs();
+    _mesh.setupFiniteVolumeMeshData();
```
Comment (lindsayad) [inline framework/src/problems/DisplacedProblem.C:287]: Is it necessary to call both `markFiniteVolumeInfoDirty` and `setupFiniteVolumeMeshData`? Will the latter call early return if the info is not dirty?

File: test/src/vectorpostprocessors/TestElemInfo.C [hunk 39]
```diff
@@ -0,0 +1,80 @@
...
+    const auto & elem_info = ei_pair.second;
+    _elem_id.push_back(elem_info->elem()->id());
+    _elem_volume.push_back(elem_info->volume() * elem_info->coordFactor());
+
+    const Point & centroid = elem_info->centroid();
+    _cx.push_back(centroid(0));
+    _cy.push_back(centroid(1));
+    _cz.push_back(centroid(2));
+
+    for (unsigned int l = 0; l < _vars.size(); ++l)
```
Comment (lindsayad) [inline test/src/vectorpostprocessors/TestElemInfo.C:64]: Could use `make_range`

File: test/tests/mesh/elem_info/elem_info.i [hunk 40]
```diff
@@ -0,0 +1,51 @@
...
+    ymax = 3
+    ny = 3
+  []
+  allow_renumbering = false
+[]
+
+[Problem]
+  kernel_coverage_check = false
+  solve = false
+
```
Comment (lindsayad) [inline test/tests/mesh/elem_info/elem_info.i:18]: Remove new line

File: test/tests/mesh/elem_info/tests [hunk 41]
```diff
@@ -0,0 +1,35 @@
...
+      detail = "quadrilateral elements,"
+    []
+    [rz-quad]
+      type = 'CSVDiff'
+      input = 'elem_info.i'
+      max_parallel = 1
+      csvdiff = 'cylindrical_quads_elem_info_0001.csv'
+      recover = false
+      cli_args = "Outputs/file_base='cylindrical_quads' Mesh/coord_type='RZ' Mesh/rz_coord_blocks='0' Mesh/rz_coord_origins='0 -2 0' Mesh/rz_coord_directions='1 0 0'"
+      detail = "quadrilateral elements in cylindrical coordinate system, and"
```
Comment (lindsayad) [inline test/tests/mesh/elem_info/tests:23]: "in cylindrical" -> "in a cylindrical"

File: framework/src/problems/DisplacedProblem.C (hunk 38 above)
Comment (grmnptr) [reply framework/src/problems/DisplacedProblem.C:287]: No, you are right there is no early return anymore. Will fix this.

**Review by lindsayad — APPROVED** [review]

**Comment by grmnptr** [comment]
I am going to shorten some of the segregated solver tests to make sure they don't time out time-to-time in debug.

**Comment by grmnptr** [comment]
I am also going to run a profiler to see if I am messing this up with the caching.

**Comment by grmnptr** [comment]
On a case with 10,000 cells (2D lid driven cavity) and 4 DoFs per cell (u,v,p,T):
The profiles says that the additional caching is less than 0.05% of the runtime.
The memory contribution from the `buildFiniteVolumeInfo` is 4%
The memory contribution from the `cacheFaceInfoOwnership` is 1.4%
The memory contribution from the `cacheFVElementalDoFs` is 0.67%

For comparison the dof-lookup right now (the initial condition does not need it anymore, but the residual evaluations still operate on it) is about 10% of the total runtime. Admittedly, the segregated  solver is very residual evaluation intense so in the monolithic solver it is probably less.

**Comment by GiudGiud** [comment]
We can still use FE ICs on FV variables right?

**Comment by grmnptr** [comment]
Not anymore.

**Comment by grmnptr** [comment]
Hmmm there might be one error that I am missing.

**Comment by GiudGiud** [comment]
It would be good if we could. 
We just lost the ability to restart from solution files if we cant.

**Comment by grmnptr** [comment]
I added a test from exodus restart with fvics. It is not icdependent. Maybe
we are talking about different things.

**Comment by GiudGiud** [comment]
Yeah solution fles are closer to checkpoints than than Exodus files

**Comment by grmnptr** [comment]
Okay I'll take a look tomorrow. The be honest right now we are erroring if the FVIC is defined for a fevariable but not the other way around. I'll check how much of a hickup it would be to restart fv variables without feics.

**Comment by grmnptr** [comment]
Also, could you point me to a test where we restart fv variables from a solution file? Thanks.

**Comment by GiudGiud** [comment]
well there is not, no guarantee that it works right now 
See SolutionIC object

**Comment by grmnptr** [comment]
Ok I'll give it a shot tomorrow but I feel like exodus and checkpoint restarts should be enough. We would loose some other ICs too, you are right. But at this point 100% of the use cases are covered by the added parts in this PR. Unless the vtb has special ics. Maybe I won't add the error to restrict ics to fevariables. The reason why i wanted to separate these things is that the linear fv variable wont have any of the data members and wont use the assembly which are essential parts of the ic-loop for fes.
